File: apps/gpt-sites/src/runtime/review/explainer.ts
import type { Explanation, Finding } from "@/src/contracts/review";

const BOUNDARY =
  "This explanation restates deterministic evidence. It does not change the finding, certify compliance, or replace professional review.";

export function explainFinding(finding: Finding): Explanation {
  let whyItMatters: string;
  if (finding.status === "FAIL") {
    whyItMatters =
      "The available explicit evidence is sufficient for comparison and does not satisfy the configured project rule.";
  } else if (finding.status === "REVIEW") {
    whyItMatters =
      "The model evidence is missing, ambiguous, contradictory, or only a proxy, so a defensible pass/fail decision would overstate certainty.";
  } else {
    whyItMatters =
      "The available explicit evidence satisfies this configured project rule for the specific element checked.";
  }

  return {
    summary: finding.message,
    why_it_matters: whyItMatters,
    next_step: finding.recommendation,
    boundary: BOUNDARY,
  };
}

export function attachExplanations(findings: Finding[]): Finding[] {
  return findings.map((finding) => ({
    ...finding,
    explanation: explainFinding(finding),
  }));
}

File: src/bim_review_agent/application/explainer.py
"""Bounded deterministic explanations for completed findings."""

from __future__ import annotations

from bim_review_agent.domain.models import Explanation, Finding, FindingStatus

_BOUNDARY = (
    "This explanation restates deterministic evidence. It does not change the finding, "
    "certify compliance, or replace professional review."
)


def explain_finding(finding: Finding) -> Explanation:
    if finding.status is FindingStatus.FAIL:
        why = (
            "The available explicit evidence is sufficient for comparison and does not satisfy "
            "the configured project rule."
        )
    elif finding.status is FindingStatus.REVIEW:
        why = (
            "The model evidence is missing, ambiguous, contradictory, or only a proxy, so a "
            "defensible pass/fail decision would overstate certainty."
        )
    else:
        why = (
            "The available explicit evidence satisfies this configured project rule for the "
            "specific element checked."
        )
    return Explanation(
        summary=finding.message,
        why_it_matters=why,
        next_step=finding.recommendation,
        boundary=_BOUNDARY,
    )


def attach_explanations(findings: list[Finding]) -> list[Finding]:
    return [item.model_copy(update={"explanation": explain_finding(item)}) for item in findings]
